File: tests/fixtures/csharp-to-cpp-port/sample-app/Services/Logger.cs
using System;
using System.IO;

namespace SampleApp.Services
{
    public sealed class Logger : IDisposable
    {
        private readonly StreamWriter _writer;
        private int _count;

        public Logger(string path)
        {
            if (!string.IsNullOrEmpty(path))
            {
                _writer = new StreamWriter(path, false);
            }
        }

        public int Count => _count;

        public void Log(string message)
        {
            string line = string.Format("LOG {0}: {1}", _count, message);
            _count++;
            if (_writer != null)
            {
                _writer.WriteLine(line);
            }
            else
            {
                Console.WriteLine(line);
            }
        }

        public void Dispose()
        {
            if (_writer != null)
            {
                _writer.Flush();
                _writer.Dispose();
            }
        }
    }
}

File: tests/fixtures/csharp-to-cpp-port/expected-cpp/Services/Logger.cpp
// Ported from Services/Logger.cs
#include "Services/Logger.h"

#include <iostream>

#include "PortSupport.h"

namespace SampleApp::Services {

Logger::Logger(const std::wstring& path)
{
    if (!path.empty()) {
        writer_ = PortSupport::OpenTextFile(path);
        toFile_ = true;
    }
}

Logger::~Logger() noexcept
{
    // C# Dispose(): flush + dispose the StreamWriter.
    if (toFile_) {
        writer_.flush();
        writer_.close();
    }
}

void Logger::Log(const std::wstring& message)
{
    // string.Format("LOG {0}: {1}", _count, message)
    const std::wstring line = L"LOG " + PortSupport::ToWString(count_) + L": " + message;
    count_++;
    if (toFile_) {
        PortSupport::WriteLine(writer_, line);
    } else {
        std::wcout << line << L'\n';
    }
}

} // namespace SampleApp::Services
